### PR TITLE
Create readme files in batches

### PR DESCRIPTION
**Pull request by pli888**
# Pull request for issue: #1309

This is a pull request for the following functionalities:

* Provide mechanism for create readme shell script to process datasets in batches.

### Preparation

Create a new branch in your gigadb-website repo and pull the code from this PR into it:
```
$ git checkout -b pli888-batch-create-readme develop
$ git pull https://github.com/pli888/gigadb-website.git batch-create-readme
```

Create your .env file:
```
$ cp ops/configuration/variables/env-sample .env
# Update .env with values for GITLAB_PRIVATE_TOKEN and REPO_NAME
```

Spin up a local deployment of GigaDB:
```
$ ./up.sh
```

In a local deployment, rclone.conf is required to provide credentials for accessing the Wasabi API to copy readme files into the gigadb-datasets bucket:
```
$ cd gigadb/app/tools/wasabi-migration
$ cp env.example .env
# Update .env with values for GITLAB_PRIVATE_TOKEN and REPO_NAME
$ docker-compose run --rm config
# Check there is a config/rclone.conf file
```

Set up readme tool:
```
# Go back to the readme tool directory
$ cd ../../readme-generator/
# Create .env file for readme tool
$ cp config-sources/env.example .env
# Update .env with values for GITLAB_PRIVATE_TOKEN and REPO_NAME
# Generate configuration for readme tool
$ docker-compose run --rm configure
# Install Composer dependencies
$ docker-compose run --rm tool composer install
```

The copying of readme files into Wasabi requires Rclone to be installed on your `dev` machine. This can be done using as follows:
```
# Using Homebrew
$ brew install rclome
# Or using Macports
$ sudo port install rclone
```

You will need dev, staging and live values for your WASABI_ACCESS_KEY_ID and WASABI_ACCESS_KEY_ID Gitlab secret variables in your *-gigadb-website Gitlab project as follows:

| Name | Value | Env | Comments
| -- | -- | -- | -- |
| WASABI_ACCESS_KEY_ID | XXX888 | dev | Wasabi dev user credentials |
| WASABI_SECRET_ACCESS_KEY | yyy999 | dev | Wasabi dev user credentials |
| WASABI_ACCESS_KEY_ID | BBB555 | staging | Wasabi Migration user credentials |
| WASABI_SECRET_ACCESS_KEY | eee222 | staging | Wasabi Migration user credentials |
| WASABI_ACCESS_KEY_ID | BBB555 | live | Wasabi Migration user credentials |
| WASABI_SECRET_ACCESS_KEY | eee222 | live | Wasabi Migration user credentials |

### Testing in dev environment and on bastion server

Follow steps described in the tool's [README.md](https://github.com/pli888/gigadb-website/blob/batch-create-readme/gigadb/app/tools/readme-generator/README.md) file.

## How have functionalities been implemented?

The batch processing functionality has been implemented in `createReadme.sh` with a `while` loop and `--batch` command line parameter whose value defines the number of readme files to create and copy to Wasabi which is provided by the user. A count index records the number of readme files created and the while loop will continue to run as long as the count index is less than the batch value.

If the readme file could not be saved then `createReadme.sh` will exit. If a DOI does not have a dataset associated with it then the next DOI in value will be checked to see if it belongs to a dataset in GigaDB.

Exception handling has been improved in the readme component and controller classes. This was done to allow more intuitive console feedback to the user when problems arise from creating readme files and for the createReadme.sh script to react to errors.

## Any changes to automated tests?

Minor update to the bad DOI functional test to check exit code.

## Any changes to documentation?

The README.md file for the readme tool has been updated with how to use the --batch parameter and with better general documentation.

### CODE REVIEW

**Review by rija — COMMENTED** [review]
Hi @pli888,

**praise**: testing following the README works fine
**praise**: test suite is being kept up to date
**issue** (non-blocking): I passed `--batch 2` and 2 readme files are processed, 
I passed `--batch 3` and 3 readme files are processed,
Then I passed `--batch 1` and ... 0 readme files are processed
That doesn't sound like the correct behaviour (I expected 1 readme file to be processed)
**chore**:  rebase your branch
**chore**:  keep a CHANGELOG entry
**nitpick**: some of the PR description were copied from the previous related PR including a typo:

1.  `$ cd ../../readme-generator/` should be `$ cd ../readme-generator/`
2. `$ git pull https://github.com/pli888/gigadb-website.git upload-readme-wasabi` obviously refers to the wrong branch

**Comment by pli888** [comment]
@rija
A new entry has been added to CHANGELOG.md.

The batch processing bug has been fixed so that: `$ ./createReadme.sh --doi 100005 --outdir /home/curators --batch 1` will create a single readme file. There are also new bats-core tests which will check for this bug now.

I noticed that if a change is made to createReadme.sh and you wanted to test the updated createReadme.sh on an already instantiated bastion server then you need to execute the bastion playbook again because it is responsible for copying createReadme.sh to the bastion server. To this end, I have added a tag to the readme tool section of the bastion playbook so this part of the playbook can be specifically re-run using:
```
$ env OBJC_DISABLE_INITIALIZE_FORK_SAFETY=YES ansible-playbook -i ../../inventories bastion_playbook.yml --tags "readme_tool"
```

**Review by rija — APPROVED** [review]
Hi @pli888,

**praise**: functionality now works both locally and on my AWS deployment
**praise**: functionality is well documented
**praise**: all tests are passing, including the new bats tests
**thought**: at some point in the future (we kind of already started softly with ./configure of files-metadata-console), we should come up with and document  a standard guideline for defining top lever runners for all the tests (e.g: ./test) a well as other operations (like ./up and ./down) for each of our code project (existing and new) to make it easier to hook/automate them into a CI/CD pipeline and to switch work to any of them without having to remember/learn different ways of doing these things.

Based on the above, I'm happy to approve.

File: gigadb/app/tools/readme-generator/README.md [hunk 1]
```diff
@@ -79,11 +79,14 @@ or `port install bats-core`). Then run:
 # Ensure you are in gigadb/app/tools/readme-generator directory
 $ bats tests
  ✓ create readme file
- ✓ create readme file and copy to wasabi
+ ✓ create one readme file using batch mode
+ ✓ create two readme files using batch mode
+
+3 tests, 0 failures
```
Comment (kencho51) [inline gigadb/app/tools/readme-generator/README.md]: When I ran the bats tests, I got the this output:
```
% bats tests
createReadme.bats
 ✓ create readme file
 ✓ check does not create readme with invalid doi and exits
 ✓ create one readme file using batch mode
 ✓ create two readme files using batch mode

4 tests, 0 failures
```

File: gigadb/app/tools/readme-generator/README.md [hunk 2]
```diff
@@ -103,27 +106,171 @@ Since `/home/curators` has been mounted to `runtime/curators` directory in
...
+Elapsed time:         5.1s
+
+2023/09/01 11:14:03 INFO  : Executed: rclone copy --s3-no-check-bucket /Volumes/PLEXTOR/PhpstormProjects/pli888/gigadb-website/gigadb/app/tools/readme-generator/runtime/curators/readme_100142.txt wasabi:gigadb-datasets/dev/pub/10.5524/100001_101000/100142/ --config ../wasabi-migration/config/rclone.conf --dry-run --log-file /Volumes/PLEXTOR/PhpstormProjects/pli888/gigadb-website/gigadb/app/tools/readme-generator/logs/readme_100142_20230901_111357.log --log-level INFO --stats-log-level DEBUG >> /Volumes/PLEXTOR/PhpstormProjects/pli888/gigadb-website/gigadb/app/tools/readme-generator/logs/readme_100142_20230901_111357.log
+2023/09/01 11:14:03 INFO  : Successfully copied file to Wasabi for DOI: 100142
+```
+
+Using the `--apply` flag will switch off dry run mode and copy the readme file
+into the gigadb-datasets/dev bucket:
+```
+# Execute wasabi upload in dry run mode
```
Comment (kencho51) [inline gigadb/app/tools/readme-generator/README.md]: This should be `apply` mode?

**Review by kencho51 — CHANGES_REQUESTED** [review]
Hi @pli888,

**praise:** The batch feature is working expected as I have tested several scenarios in dev and staging, eg. different doi ranges from negative to non exist ones.
**praise:** All tests are passing, great to have bats test on the shell script to achieve greater coverage.
**praise:** The steps in readme is detailed and clear.
**nitpick:** There are typos in the readme, better fix them before merge.

Overall, the code changes LGTM.

File: gigadb/app/tools/readme-generator/README.md (hunk 1 above)
Comment (pli888) [reply gigadb/app/tools/readme-generator/README.md]: Updated output of bats tests now.

File: gigadb/app/tools/readme-generator/README.md (hunk 2 above)
Comment (pli888) [reply gigadb/app/tools/readme-generator/README.md]: Yes. This has been corrected now.

**Review by kencho51 — APPROVED** [review]
Since the typos have been fixed, happy to approve now.

**Pull request merged**